### PR TITLE
First cut at enabling builds with MSVC2019 + clang-cl.exe

### LINKED ISSUE #779: Add support for clang-cl (Visual Studio 2019)

### PR DESCRIPTION
**Pull request by KineticTheory**
### Background

+ I would like to be able to use both the MS C++ compiler and the LLVM C++ compiler under Visual Studio 2019.  This PR provides a first cut at enabling this feature.  This will lead to an ability to run _clang-tidy_ inside of Visual Studio.
+ WIP: requires these PRs to be merged first:
  + [x] #780
  + [x] #781
  + [x] #782
  + [x] Rebase after above are merged.

### Purpose of Pull Request

* Fixes #779 
* Related to #780, #781, #782

### Description of changes

+ Add a LLVM build profile to `CMakeSettings.json`.
+ Add a compiler flag definition file `windows-clang.cmake` and source this file when setting up the compiler.
+ Disable a debug print that was accidentally left enabled in a previous commit.  This debug print would only show up when running CMake that targeted a "Visual Studio" project.
+ Currently, the compiler flags are the same as those found in `windows-cl.cmake`, but I expect them to diverge soon.
+ Fix a logic bug in `ds++/CMakeLists.txt` that prevented `clang-cl.exe` from linking against the _winsock_ library.
+ For now the compiler warning level us set to 0 (do not report warnings).

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: config/component_macros.cmake [hunk 1]
```diff
@@ -712,8 +712,8 @@ function( copy_dll_link_libraries_to_build_dir target )
           get_target_property( isimp ${lib} IMPORTED )
           if(isimp)
             get_target_property( link_libs3 ${lib} INTERFACE_LINK_LIBRARIES)
-          else()
-            get_target_property( link_libs2 ${lib} LINK_LIBRARIES )
+          #else()
+          #  get_target_property( link_libs2 ${lib} LINK_LIBRARIES )
           endif()
           list( APPEND link_libs ${link_libs2} )
```
Comment (timmah) [inline config/component_macros.cmake]: Is `link_libs2` defined at this point?

Comment (KineticTheory) [reply config/component_macros.cmake]: @timmah Let me rebase before you review this PR.  This above change was necessary for cmake-3.17.  But I might remove it from this PR.

**Comment by keadyk** [comment]
@KineticTheory is a clang-tidy build going to join our nightlies at some point? (I assume initial attempts will throw a bajillion warnings and require some cleanup, so this might be a long term goal) :D

**Comment by KineticTheory** [comment]
> @KineticTheory is a clang-tidy build going to join our nightlies at some point? (I assume initial attempts will throw a bajillion warnings and require some cleanup, so this might be a long term goal) :D

The plan is to add clang-tidy to the regression system.  If you look at `src\CMakeLists.txt` you will see the logic for only running it on a subset (only `ds++`) of Draco.

**Comment by timmah** [comment]
If you think of any custom checks that clang-tidy doesn't cover, let me know. Clang/libtooling is a bit of a hobby.

**Comment by KineticTheory** [comment]
> If you think of any custom checks that clang-tidy doesn't cover, let me know. Clang/libtooling is a bit of a hobby.

Sure thing.  Right now I'm just running the *modernize* checks. See `config/comilerEnv.cmake` and grep for `CLANG_TIDY_CHECKS`.

File: config/windows-clang.cmake [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+# Useful reference information:
+# https://docs.microsoft.com/en-us/cpp/intrinsics/cpuid-cpuidex?view=vs-2017
+
+#
+# Sanity Checks
+#
+if( NOT ${CMAKE_GENERATOR} MATCHES "Visual Studio" AND
+    NOT ${CMAKE_GENERATOR} MATCHES  "NMake Makefiles" )
+  message( FATAL_ERROR
+  "config/windows-cl.cmake must be taught to build for this compiler "
```
Comment (timmah) [inline config/windows-clang.cmake]: Should be "config/windows-clang.cmake must..." ?

File: config/windows-clang.cmake [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  # Alternative for per-directory, or per-target specific flags:
+  # add_compile_options("$<$<CXX_COMPILER_ID:MSVC>:/MP>")
+
+  # Notes on options:
+  # - /wd 4251 disable warning #4251: 'identifier' : class 'type' needs to have
+  #            dll-interface to be used by clients of class 'type2'
+  # - /arch:[SSE|SSE2|AVX|AVX2|IA32]
+  # - /W[1234] Warning levels. Draco is currently using /W4 (see
+  #            src/CMakeLists.txt), but clients default to /W2.
+  # - /std:c++14 (should be added by CMake in compilerEnv.cmake)
```
Comment (timmah) [inline config/windows-clang.cmake]: Update comment to "/W0"?

**Comment by timmah** [comment]
Couple notes on comments in `windows-clang.cmake`, otherwise looks good.

File: config/windows-clang.cmake (hunk 3 above)
Comment (KineticTheory) [reply config/windows-clang.cmake]: Fixed in aeb6827.

File: config/windows-clang.cmake (hunk 2 above)
Comment (KineticTheory) [reply config/windows-clang.cmake]: Fixed in aeb6827.

**Comment by timmah** [comment]
Nice. Are you planning any more changes? This looks good to go.

**Comment by KineticTheory** [comment]
> Nice. Are you planning any more changes? This looks good to go.

No more changes for this PR, although I have a follow-on with more clang-tidy specific changes.  Please merge at your convenience.

**Pull request merged**